fixes #53 combinatorial memory usage in `dict` arg

richpsharp opened the pull request:
This PR fixes #53 by constructing a scrubbed dictionary rather than an incorrect series of dictionary value tuples. Added a test that would have exposed this issue and updates history.

dcdenu4 commented on taskgraph/Task.py:
```diff
@@ -1533,9 +1532,11 @@ def _filter_non_files(
                     value, keep_list, ignore_list, keep_directories):
-                yield (value, filter_value)
+                scrubbed_dict[key] = filter_value
+            yield scrubbed_dict
```
Should this `yield` be inside the for loop? If it's not I think it will overwrite nested values. Passing in:
`{0: {'one': 0, 'two': 1, 'three': 2}, 1: {'one': 1, 'two': 2, 'three': 3}, 2: {'one': 2, 'two': 3, 'three': 4}}`
gives me a generator that produces:
```
{0: {'three': 2}}
{1: {'three': 3}}
{2: {'three': 4}}
```
Where as kicking it inside the loop gives me:
```
{0: {'one': 0}}
{0: {'two': 1}}
{0: {'three': 2}}
{1: {'one': 1}}
{1: {'two': 2}}
{1: {'three': 3}}
{2: {'one': 2}}
{2: {'two': 3}}
{2: {'three': 4}}
```

richpsharp replied:
Gosh thanks for spotting this, there was a lot more going on there even than just that.

dcdenu4 reviewed (changes requested):
Hey @richpsharp thanks for looking into this so quickly. I just had one comment about where the `yield` should be. On that topic, the `_filter_non_files` functions says it returns the `base_value`, but really it returns a generator no? Sorry, just trying to figure out what I should be expecting back form this function. Thanks!

richpsharp commented:
> Hey @richpsharp thanks for looking into this so quickly. I just had one comment about where the `yield` should be. On that topic, the `_filter_non_files` functions says it returns the `base_value`, but really it returns a generator no? Sorry, just trying to figure out what I should be expecting back form this function. Thanks!

Hi @dcdenu4, thanks for doing a careful test on this. The result you spotted showed that `_filter_non_files` was not behaving as expected at all. I simplified the implementation of `_filter_non_files` and fixed an additional bug I found in creating directories. That's all rolled into a more extensive test suite now. 

p.s. I love that idea from Pragmatic Programmer that if you find a bug it means your test suite wasn't good enough so add a test and then fix it. I did that with the sample you posted above fyi.

dcdenu4 commented on tests/test_task.py:
```diff
@@ -1494,6 +1495,85 @@ def test_terminate_log(self):
+        from taskgraph.Task import _normalize_path
+
+        # Test a passthrough
+        test_dict = {
+            0: {'one': 0, 'two': 1, 'three': 2},
+            1: {'one': 1, 'two': 2, 'three': 3},
+            2: {'one': 2, 'two': 3, 'three': 4}}
+        self.assertEqual(
+            test_dict, _filter_non_files(test_dict, [], [], False))
```
Nice!

dcdenu4 reviewed (approved):
Looks great! I just finished a read through the updated Pragmatic Programmer. So many good things in there, like: "A test is the first user of your code". What a nugget.

dcdenu4 commented:
> p.s. I love that idea from Pragmatic Programmer that if you find a bug it means your test suite wasn't good enough so add a test and then fix it. I did that with the sample you posted above fyi.

Yes, I remember reading this just a few days ago. Don't assume your patch fixed everything and it won't happen again, because it will! Gotta get that test in there.
